src/loader/mod.rs

```diff
@@ -211,20 +215,24 @@ pub fn load_cmdline<M: GuestMemory>(
     guest_addr: GuestAddress,
     cmdline: &Cmdline,
 ) -> Result<()> {
-    let len = cmdline.as_str().len();
-    if len == 0 {
-        return Ok(());
-    }
+    // We need a null terminated string because that's what the Linux
+    // kernel expects when parsing the command line:
+    // https://elixir.bootlin.com/linux/v5.10.139/source/kernel/params.c#L179
+    let cmdline_string = cmdline
+        .as_cstring()
+        .map_err(|_| Error::InvalidCommandLine)?;
+
+    let cmdline_bytes = cmdline_string.as_bytes_with_nul();
 
     let end = guest_addr
-        .checked_add(len as u64 + 1)
-        .ok_or(Error::CommandLineOverflow)?; // Extra for null termination.
+        .checked_add((cmdline_bytes.len() - 1) as u64)
```

Comment:
nit: I would leave a comment here explaining that the subtraction can not underflow because cmdline_bytes has at least one byte (the nul terminator),

Reply:
adding that as well